### PR TITLE
add some log for non using detached tables

### PR DESCRIPTION
**Pull request by k-morozov**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

I suggest to add some information in log when we call dtor in removing non using detached tables. With it we can understand how many time we spend in long detach for clear non using tables.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit bdd8bcc0d9b68474ca10df52772babc5aa1a20d4 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64992/bdd8bcc0d9b68474ca10df52772babc5aa1a20d4/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by k-morozov** [comment]
@antaljanosbenjamin It seems some checks pinding.

File: src/Databases/DatabaseAtomic.cpp
Comment (antaljanosbenjamin) [inline src/Databases/DatabaseAtomic.cpp:119]:
```suggestion
        LOG_DEBUG(log, "Finished removing not used detached tables");
```

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (antaljanosbenjamin) [inline src/Storages/StorageReplicatedMergeTree.cpp:20]:
```suggestion
#include <Common/logger_useful.h>
#include <Common/Macros.h>
#include <Common/MemoryTracker.h>
#include <Common/ProfileEventsScope.h>
#include <Common/StringUtils.h>
#include <Common/ZooKeeper/KeeperException.h>
#include <Common/ZooKeeper/Types.h>
#include <Common/escapeForFileName.h>
#include <Common/formatReadable.h>
#include <Common/noexcept_scope.h>
```

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (antaljanosbenjamin) [inline src/Storages/StorageReplicatedMergeTree.cpp:5290]:
```suggestion
            LOG_TRACE(log, "The attach thread is shutdown");
```

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (antaljanosbenjamin) [inline src/Storages/StorageReplicatedMergeTree.cpp:5306]:
```suggestion
    LOG_TRACE(log, "Finished preparing for shutdown");
```

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (antaljanosbenjamin) [inline src/Storages/StorageReplicatedMergeTree.cpp:5346]:
```suggestion
    LOG_TRACE(log, "Shutdown started");
```
Doesn't the log already contain the table name and uuid?

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (antaljanosbenjamin) [inline src/Storages/StorageReplicatedMergeTree.cpp:5389]:
```suggestion
    LOG_TRACE(log, "Shutdown finished, table={}", storage_name);
```
```suggestion
    LOG_TRACE(log, "Shutdown finished");
```

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (k-morozov) [reply src/Storages/StorageReplicatedMergeTree.cpp:5346]: ok

File: src/Storages/StorageReplicatedMergeTree.cpp
Comment (k-morozov) [reply src/Storages/StorageReplicatedMergeTree.cpp]: We cann't remove FailPoint and ThreadFuzzer.

**Review by antaljanosbenjamin — APPROVED** [review]

**Comment by antaljanosbenjamin** [comment]
@k-morozov please do not try to skip tests even for trivial changes. Our tests sometimes quite complex, even simple looking changes can break something.

**Comment by antaljanosbenjamin** [comment]
```
Jun 17 13:23:12 /build/src/Storages/StorageReplicatedMergeTree.cpp:5344:16: error: unused local variable 'storage_name' of type 'const String' (aka 'const basic_string<char, char_traits<char>, allocator<char>>') [bugprone-unused-local-non-trivial-variable,-warnings-as-errors]
Jun 17 13:23:12  5344 |     const auto storage_name = getStorageID().getNameForLogs();
Jun 17 13:23:12       |                ^
```

Could you please fix this [clang-tidy warning](https://s3.amazonaws.com/clickhouse-builds/PRs/64992/f7eac01b822c94184a16dfda1685d95f05c5cc8a/binary_tidy/build_log.log)?

**Review by antaljanosbenjamin — APPROVED** [review]

**Comment by antaljanosbenjamin** [comment]
So it seems both public and private CI is in much better shape. Could you please merge a recent master into this PR?

**Comment by k-morozov** [comment]
> So it seems both public and private CI is in much better shape. Could you please merge a recent master into this PR?

yea, sure.

**Comment by antaljanosbenjamin** [comment]
@k-morozov I know I am dragging this PR a bit longer than it should be, but please be patient. Don't merge master just to try to make the tests green. I will check the failures soon.

File: src/Databases/DatabaseAtomic.cpp
Comment (antaljanosbenjamin) [inline src/Databases/DatabaseAtomic.cpp:121]: I think this is a potential deadlock, see [here](https://s3.amazonaws.com/clickhouse-test-reports/64992/4a460760a7a55c3caeaf5537ac24437f90d9640d/integration_tests__tsan__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log).

Basically before your changes we first released the lock and then cleaned up the not used tables, while after your changes we clean up the tables while holding the lock. Maybe it is better to introduce a scope for the lock and do this outside of that scope.

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]

File: src/Databases/DatabaseAtomic.cpp
Comment (antaljanosbenjamin) [reply src/Databases/DatabaseAtomic.cpp:121]: And maybe put some comment into the code to be obvious why it is necessary.

Comment (k-morozov) [reply src/Databases/DatabaseAtomic.cpp:126]: Unfortently, Im not clear understand logic within dtors. However you right about diffrent order after my changes. I will fix it and add comments about the order. Thank you.

Comment (antaljanosbenjamin) [reply src/Databases/DatabaseAtomic.cpp:126]: Maybe let me try to explain it. It is a good example for the future too.

We have to mutexes:
1. `IDatabase::mutex` (let's call it `DM`), which is used to attach/detach tables etc.
2. `BackgroundSchedulePoolTaskInfo::exec_mutex` (let's call it `BM`), which is used when we do some operations on the task, e.g. activate or deactivate.

To execute a background task, we have to lock `BM`. In the background task of `StorageKafka`, we have to retrieve some table ids, so we call `IDatabase::tryGetTable`, which internally locks `DM`. Thus in this path `BM` is locked before `DM`. 

The dtor of `StorageKafka` destroys the `BackgroundSchedulePoolTaskHolder`s, which calls `BackgroundSchedulePoolTaskInfo::deactivate`, which locks `BM`. With you "old" changes, `~StorageKafka` is called while holding `DM`, thus in this way `DM` is locked before `BM`.

This means if we try to destroy a StorageKafka while one of its background tasks are trying to retrieve a table id, we would get into a deadlock.

File: src/Databases/DatabaseAtomic.cpp
Comment (antaljanosbenjamin) [inline src/Databases/DatabaseAtomic.cpp:110]:
```suggestion
    // it is important to call destructors of not_in_use without
    // locked mutex for avoid potential deadlock.
```

**Review by antaljanosbenjamin — APPROVED** [review]

**Comment by antaljanosbenjamin** [comment]
- `Performance Comparison (release) [4/4]` https://github.com/ClickHouse/ClickHouse/pull/66342, also other failures are unlikely to be connected to this PR.
 - `01053_ssd_dictionary` - https://github.com/ClickHouse/ClickHouse/issues/66193
 - `01276_alter_rename_column_materialized_expr` is a random `Access Denied` from minio, still investigating what happened there, but very unlikely that it is caused by this PR.

**Pull request merged**
